packages/forms/test/reactive_integration_spec.ts

```diff
@@ -4358,6 +4358,369 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+               // [formGroup] *ngIf
+
+               const group = new FormGroup({});
+
+               @Component({
+                 selector: 'app',
+                 template: `
+                    <div *ngIf="visibleA" [formGroup]="group" app-async-validator></div>
+                    <div *ngIf="visibleB" [formGroup]="group" app-async-validator></div>
+                  `
+               })
+               class App {
+                 visibleA = true;
+                 visibleB = true;
+                 group = group;
+               }
+
+               const fixture = initCleanupTest(App, AppAsyncValidator);
+
+               const div = fixture.debugElement.query(By.css('div'));
+               const appAsyncValidator = div.injector.get(AppAsyncValidator);
+
+               fixture.componentInstance.visibleA = false;
+               fixture.detectChanges();
+
+               // It was called forcely from the `updateValueAndValidity()` method.
+               expect(appAsyncValidator.unsubscribeCalledTimes).toEqual(1);
+
+               fixture.componentInstance.visibleB = false;
+               fixture.detectChanges();
+
+               // Previously, it would not have unsubscribed from observables after
+               // the directive was destroyed.
+               // Let's ensure that the observable is unsubscribed when the directive is destroyed.
+               expect(appAsyncValidator.unsubscribeCalledTimes).toEqual(2);
+             });
+
+          it('should not unsubscribe from validator when the FormGroup has own async validator function',
+             () => {
+               // Scenario:
+               // ---------
+               // [formGroup] *ngIf
+
+               let ownAsyncValidatorUnsubscribeCalledTimes = 0;
+
+               const group = new FormGroup({}, {
+                 asyncValidators: () => new Observable(() => {
+                   return () => {
+                     ownAsyncValidatorUnsubscribeCalledTimes++;
+                   };
+                 })
+               });
+
+               @Component({
+                 selector: 'app',
+                 template: '<div *ngIf="visible" [formGroup]="group" app-async-validator></div>'
+               })
+               class App {
+                 visible = true;
+                 group = group;
+               }
+
+               const fixture = initCleanupTest(App, AppAsyncValidator);
+
+               const div = fixture.debugElement.query(By.css('div'));
+               const appAsyncValidator = div.injector.get(AppAsyncValidator);
+
+               fixture.componentInstance.visible = false;
+               fixture.detectChanges();
+
+               // It was called forcely from the `updateValueAndValidity()` method.
+               expect(ownAsyncValidatorUnsubscribeCalledTimes).toEqual(1);
+
+               // It doesn't unsubscribe from the observable since FormGroup has its
+               // own validator function.
```

Comment:
Note: I've created a new ticket https://github.com/angular/angular/issues/40943 to capture the proposal to make subscriptions more granular. You can refer to that ticket in this comment (and in other comments in your PR) to provide additional context, for example:
```suggestion
               // It doesn't unsubscribe from the observable since FormGroup has its
               // own validator function (which gets combined into a single validator function in AbstractControl),
               // so we can not unsubscribe without stopping own async validator.
               // See https://github.com/angular/angular/issues/40943 for additional info.
```

Reply:
I also think that it'd be great to split tests into 2 sub-groups: where we verify that subscription is removed and where we keep subscription due to internal data structures, something like this:
```
  describe('complete cleanup of async validation subscription', () => { ... });

  // Tests below should be revisited as a part of https://github.com/angular/angular/issues/40943,
  // currently we just verify that we are not cancelling async validation altogether if there are own control validators.
  describe('partial cleanup (async validation subscription retained)', () => { ... });
```